share/wake/lib/core/json.wake

```diff
@@ -169,3 +169,143 @@ export def x ==/ y = match x y
+# other implementation may be parsing generated output, `normalizeJSONCompat`
+# may be used to target that lowest common denominator.
+#
+# Note that the relevant normalization is applied to each member of a `JArray`
+# or `JObject` *before* the enclosing container. This means that if some
+# normalization fails anywhere in the JSON tree the entire call will fail, even
+# if the specific branch causing the failure would later have been pruned.
+export def normalizeJSONWith (fmt: JSONNormalize) (json: JValue): Result JValue Error =
+    def normalizeJSON (path: String) = match _
+        JString a =
+            getJSONNormalizeString fmt a
+            | rmap JString
+            | addErrorContext path
+        JInteger a =
+            getJSONNormalizeInteger fmt a
+            | rmap JInteger
+            | addErrorContext path
+        JDouble a =
+            getJSONNormalizeDouble fmt a
+            | rmap JDouble
+            | addErrorContext path
+        JBoolean a =
+            getJSONNormalizeBoolean fmt a
+            | rmap JBoolean
+            | addErrorContext path
+        JNull =
+            Pass JNull
+        JObject a =
+            def normalizeJObjectValue entry =
+                def Pair key value =
+                    entry
+                normalizeJSON "{path}/{key}" value
+                | rmap (Pair key)
+            def normalized =
+                map normalizeJObjectValue a
+            require Pass children =
+                findFail normalized
+            children
+            | getJSONNormalizeObject fmt
+            | rmap JObject
+            | addErrorContext path
+        JArray a =
+            def normalized =
+                zip (seq a.len) a
+                | map (\(Pair i x) normalizeJSON "{path}/[{format i}]" x)
+            require Pass children =
+                findFail normalized
+            children
+            | getJSONNormalizeArray fmt
+            | rmap JArray
+            | addErrorContext path
+    normalizeJSON "" json
+
+# The rules by which `normalizeJSONWith` will simplify JSON values.
+# If any particular function is unable to operate on every input value (e.g. a
+# `JObject` contains duplicate keys of types which cannot be combined), that
+# rule may return a `Fail` which halts the broader processing.
+export tuple JSONNormalize =
+    String: String => Result String Error
+    Integer: Integer => Result Integer Error
+    Double: Double => Result Double Error
+    Boolean: Boolean => Result Boolean Error
+    Object: List (Pair String JValue) => Result (List (Pair String JValue)) Error
+    Array: List JValue => Result (List JValue) Error
+
+# A basic normalization ruleset which returns inputs unchanged.
+export def normalizeJSONIdentity: JSONNormalize =
+    JSONNormalize Pass Pass Pass Pass Pass Pass
+
+# Target the minimum JSON language described by the specifications.
+export def normalizeJSONCompat: JSONNormalize =
```

Comment:
> Is normal json json4? If so, I'd include a '4' in the name and also add a 'json5' version, which does not mess around with Doubles.

"Compat" (renamed from "Basic") JSON is JSON according to the standards published by IETF/ECMA/ISO. JSON5 is a (presumably widespread but unofficial) effort to bring it in line with ECMAScript 5 (it's not an incremental version). I'm not convinced we need a `normalizeJSON5` object predefined; if it's critical for anyone's workflow, they can simply bypass the `JDouble` normalization:

```
def normalizeJSON5: JSONNormalize =
    normalizeJSONCompat
    | setJSONNormalizeDouble Pass
```

Reply:
So, I was misled into believing JSON5 was an actual standard. I agree we should default to the 'real' JSON and normalize to that by default. Don't bother with a JSON5 version.